libraries/chain/include/eosio/chain/webassembly/interface.hpp

```diff
@@ -0,0 +1,581 @@
+         REGISTER_CF_HOST_FUNCTION(printi128);
+         REGISTER_CF_HOST_FUNCTION(printui128);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printsf);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printdf);
+         REGISTER_CF_HOST_FUNCTION(printqf);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printn);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printhex);
+
+         // database api
+         // primary index api
+         int32_t db_store_i64(uint64_t, uint64_t, uint64_t, uint64_t, legacy_array_ptr<const char>);
+         void db_update_i64(int32_t, uint64_t, legacy_array_ptr<const char>);
+         void db_remove_i64(int32_t);
+         int32_t db_get_i64(int32_t, legacy_array_ptr<char>);
+         int32_t db_next_i64(int32_t, uint64_t&);
+         int32_t db_previous_i64(int32_t, legacy_ptr<uint64_t>);
+         int32_t db_find_i64(uint64_t, uint64_t, uint64_t, uint64_t);
+         int32_t db_lowerbound_i64(uint64_t, uint64_t, uint64_t, uint64_t);
+         int32_t db_upperbound_i64(uint64_t, uint64_t, uint64_t, uint64_t);
+         int32_t db_end_i64(uint64_t, uint64_t, uint64_t);
+
+         REGISTER_LEGACY_HOST_FUNCTION(db_store_i64);
+         REGISTER_LEGACY_HOST_FUNCTION(db_update_i64);
+         REGISTER_LEGACY_HOST_FUNCTION(db_remove_i64);
+         REGISTER_LEGACY_HOST_FUNCTION(db_get_i64);
+         REGISTER_HOST_FUNCTION(db_next_i64);
+         REGISTER_LEGACY_HOST_FUNCTION(db_previous_i64);
+         REGISTER_LEGACY_HOST_FUNCTION(db_find_i64);
+         REGISTER_LEGACY_HOST_FUNCTION(db_lowerbound_i64);
+         REGISTER_LEGACY_HOST_FUNCTION(db_upperbound_i64);
+         REGISTER_LEGACY_HOST_FUNCTION(db_end_i64);
+
+
+         // uint64_t secondary index api
+         int32_t db_idx64_store(uint64_t, uint64_t, uint64_t, uint64_t, const uint64_t&);
+         void db_idx64_update(int32_t, uint64_t, const uint64_t&);
+         void db_idx64_remove(int32_t);
+         int32_t db_idx64_find_secondary(uint64_t, uint64_t, uint64_t, legacy_ptr<const uint64_t>, legacy_ptr<uint64_t>);
+         int32_t db_idx64_find_primary(uint64_t, uint64_t, uint64_t, uint64_t&, uint64_t);
+         int32_t db_idx64_lowerbound(uint64_t, uint64_t, uint64_t, legacy_ptr<uint64_t, 8>, legacy_ptr<uint64_t, 8>);
+         int32_t db_idx64_upperbound(uint64_t, uint64_t, uint64_t, legacy_ptr<uint64_t, 8>, legacy_ptr<uint64_t, 8>);
+         int32_t db_idx64_end(uint64_t, uint64_t, uint64_t);
+         int32_t db_idx64_next(int32_t, uint64_t&);
+         int32_t db_idx64_previous(int32_t, uint64_t&);
+
+         REGISTER_HOST_FUNCTION(db_idx64_store);
+         REGISTER_HOST_FUNCTION(db_idx64_update);
+         REGISTER_LEGACY_HOST_FUNCTION(db_idx64_remove);
+         REGISTER_LEGACY_HOST_FUNCTION(db_idx64_find_secondary);
+         REGISTER_HOST_FUNCTION(db_idx64_find_primary);
+         REGISTER_LEGACY_HOST_FUNCTION(db_idx64_lowerbound);
+         REGISTER_LEGACY_HOST_FUNCTION(db_idx64_upperbound);
+         REGISTER_LEGACY_HOST_FUNCTION(db_idx64_end);
```

Comment:
Why is `db_idx64_end` legacy?

Similarly for the other db end host functions below.